packages/embark-ui/src/containers/TransactionsContainer.js

```diff
@@ -34,34 +33,56 @@ class TransactionsContainer extends Component {
   }
 
   get numberOfBlocks() {
-    const blocks = this.props.blocks;
-    return !blocks.length ? 0 : blocks[0].number + 1;
+    if (this._numberOfBlocks === undefined) {
+      const blocks = this.props.blocks;
+      this._numberOfBlocks = !blocks.length ? 0 : blocks[0].number + 1;
+    }
+    return this._numberOfBlocks;
   }
 
   get estNumberOfTxs() {
-    const blocks = this.props.blocks;
-    const numBlocksInProps = blocks.length;
-    const numTxsInPropsBlocks = blocks.reduce((txCount, block) => (
-      txCount + block.transactions.length
-    ), 0);
-    const missingNumBlocks = this.numberOfBlocks - numBlocksInProps;
-    return missingNumBlocks + numTxsInPropsBlocks;
+    if (this._estNumberOfTxs === undefined) {
+      const blocks = this.props.blocks;
+      const numBlocksInProps = blocks.length;
+      const numTxsInPropsBlocks = blocks.reduce(
+        (txCount, block) => txCount + block.transactions.length,
+        0
+      );
+      const missingNumBlocks = this.numberOfBlocks - numBlocksInProps;
+      this._estNumberOfTxs = missingNumBlocks + numTxsInPropsBlocks;
+    }
+    return this._estNumberOfTxs;
+  }
+
+  get numberOfPages() {
+    if (this._numberOfPages === undefined) {
+      this._numberOfPages = Math.ceil(
+        this.estNumberOfTxs / this.numTxsToDisplay
+      );
+    }
+    return this._numberOfPages;
   }
 
-  getNumberOfPages() {
-    return Math.ceil(this.estNumberOfTxs / this.numTxsToDisplay);
+  resetNums() {
+    this._numberOfBlocks = undefined;
+    this._estNumberOfTxs = undefined;
+    this._numberOfPages = undefined;
   }
 
   changePage(newPage) {
-    this.setState({currentPage: newPage});
-    this.props.fetchBlocksFull(
-      this.numberOfBlocks - 1 - (this.numBlocksToFetch * (newPage - 1)),
-      this.numBlocksToFetch
-    );
-    this.props.fetchTransactions((newPage * MAX_TXS) + MAX_TXS);
+    if (newPage <= 0) {
+      newPage = 1;
+    } else if (newPage > this.numberOfPages) {
+      newPage = this.numberOfPages;
+    }
+    this.setState({ currentPage: newPage });
+    const blockFrom =
+      this.numberOfBlocks - 1 - this.numBlocksToFetch * (newPage - 1);
```

Comment:
Maybe for better readability of this expression, group them using `()` ?

E.g.
```
(this.numberOfBlocks - 1) - (this.numBlocksToFetch * (newPage -1))
```
Or maybe just 

```
(this.numberOfBlocks - 1) - this.numBlocksToFetch * (newPage -1)
```
Would do the trick

Reply:
I've been using the prettier extension in vscode to auto-format code I add/change. In this case it stripped away some parens similar to what you're suggesting. At first I was surprised but then realized that order-of-operations in this expression is relatively straightforward. If you have a strong feeling that readability has suffered, I'll add the parens and not run prettier on that section of code.

Reply:
Ah, got it. No it's fine! :)